### PR TITLE
add modifications to the iptables module

### PR DESCRIPTION
**Pull request by sxpert**
add order option to insert a rule at a specific level
modify docs to make them more understandable
add proper support for new form of ! on source and destination parameters
    including an example in the docs

### CODE REVIEW

**Comment by robynbergeron** [comment]
Thanks @sxpert. @LinusU please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

**Comment by gregdek** [comment]
@LinusU This change is still pending your review; do you have time to take a look and comment? Please comment with text 'shipit' or 'needs_revision' as appropriate.

**Comment by LinusU** [comment]
I think that maybe this should be split into multiple pull request so that we can discuss it separately. We had a very large discussion already on wether to support insertion at a position or not, and as this pull request is now I would not like to land it.

The problem is that this approach is simply to fragile. If the rule already exists, it will ignore the `order` parameter and that seems like a source of very hard to track down bugs. It's important for any Ansible module to always result in the same end result.

I also think that the code quality isn't quite up to par, I'll put some specific comments on it in the diff as well...

File: system/iptables.py [hunk 1]
```diff
@@ -292,7 +295,13 @@ def check_present(iptables_path, module, params):
...
 
 def append_rule(iptables_path, module, params):
-    cmd = push_arguments(iptables_path, '-A', params)
+    if params['order'] is not None:
+        cmd = [iptables_path]
+        cmd.extend(['-t', params['table']])
+        cmd.extend(['-I', params['chain'], params['order']])
+        cmd.extend(construct_rule(params))
+    else :
+        cmd = push_arguments(iptables_path, '-A', params)
```
Comment (LinusU) [inline system/iptables.py:304]: The logic here is split very strange, this function should forward stuff to `push_arguments` in both paths.

File: system/iptables.py [hunk 2]
```diff
@@ -270,8 +272,9 @@ def construct_rule(params):
     append_match(rule, params['comment'], 'comment')
     append_param(rule, params['comment'], '--comment', False)
     append_match(rule, params['ctstate'], 'state')
-    append_csv(rule, params['ctstate'], '--state')
-    append_match(rule, params['limit'] or params['limit_burst'], 'limit')
+    if len(params['ctstate']) > 0:
```
Comment (LinusU) [inline system/iptables.py:275]: Why have you added this `if`?

File: system/iptables.py [hunk 3]
```diff
@@ -270,8 +272,9 @@ def construct_rule(params):
...
+        append_csv(rule, params['ctstate'], '--state')
+    append_match(rule, params['limit'], 'limit')
```
Comment (LinusU) [inline system/iptables.py:277]: Why did you remove `limit_burst` from this `append_match` call?

File: system/iptables.py [hunk 4]
```diff
@@ -68,6 +68,12 @@
         defined chain or any of the builtin chains: 'INPUT', 'FORWARD',
         'OUTPUT', 'PREROUTING', 'POSTROUTING', 'SECMARK', 'CONNSECMARK'"
     required: true
+  order:
+    version_added: "2.1"
+    description:
+      - "Order in which to insert the rule, in this case -I is used instead
```
Comment (LinusU) [inline system/iptables.py:74]: Trailing whitespace, also, please remove the `"`

File: system/iptables.py [hunk 5]
```diff
@@ -255,8 +249,16 @@ def append_match(rule, param, match):
...
     append_param(rule, params['protocol'], '-p', False)
-    append_param(rule, params['source'], '-s', False)
-    append_param(rule, params['destination'], '-d', False)
+    source = params['source']
+    if (source is not None) and (len(source)>0) and (source[0]=='!'):
+        rule.extent(['!'])
+        source = source[1:]
+    append_param(rule, source, '-s', False)
+    destination = params['destination']
+    if (destination is not None) and (len(destination)>0) and (destination[0]=='!'):
```
Comment (LinusU) [inline system/iptables.py:258]: This looks very crammed it should have some whitespace (`missing whitespace around operator`) and I also think that it would be good to extract this into a function.

Also, `line too long`

File: system/iptables.py [hunk 6]
```diff
@@ -292,7 +295,13 @@ def check_present(iptables_path, module, params):
 
 
 def append_rule(iptables_path, module, params):
-    cmd = push_arguments(iptables_path, '-A', params)
+    if params['order'] is not None:
+        cmd = [iptables_path]
+        cmd.extend(['-t', params['table']])
+        cmd.extend(['-I', params['chain'], params['order']])
+        cmd.extend(construct_rule(params))
+    else :
```
Comment (LinusU) [inline system/iptables.py:303]: No whitespace before `:`

**Comment by gregdek** [comment]
Thanks @sxpert for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

**Comment by gregdek** [comment]
Thanks @sxpert for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@sxpert A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@sxpert Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@sxpert A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@sxpert Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@sxpert A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@sxpert Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@sxpert A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
Closing due to contributor inactivity.
